Catch regression when sending right key when switching to tab Bootloader Options in ncurses

tinawang123 opened the pull request:
Catch regression when sending right key when switching to tab Bootloader Options in ncurses

- Related ticket: https://progress.opensuse.org/issues/124610
- Verification run: https://openqa.suse.de/tests/11038962#step/disable_boot_menu_timeout_navigating_tabs/3
https://openqa.suse.de/tests/11041255#step/disable_boot_menu_timeout_navigating_tabs/3

okurz commented on tests/installation/bootloader_settings/disable_boot_menu_timeout_testapi.pm:
```diff
@@ -0,0 +1,20 @@
+# SUSE's openQA tests
+#
+# Copyright 2021 SUSE LLC
```
```suggestion
# Copyright SUSE LLC
```

See our contribution hints for generic remarks regarding copyright

tinawang123 commented on schedule/yast/textmode/ha_textmode_skip_registration_bsc1208266.yaml:
```diff
@@ -0,0 +1,28 @@
+name: ha_textmode_skip_registration
```
For case ' ha_textmode_skip_registration ', it doesn't have default yaml file at x86_64, I need create a new one for aarch64. As I have ticket 'https://progress.opensuse.org/issues/124035' to do this task. I'd like to create default yaml file at other PR instead of this one.

tinawang123 commented:
> Please, prepare also the staging test, we should reuse the textmode test suite that we recently added (coordinate with @rakoenig in https://progress.opensuse.org/issues/124676) Once addressed feedback and verified, we will merge the PR and resolve the ticket without enabling it in job groups for staging. We should file a ticket for the enablement (but test here that all works as expected) as the bug is still not fixed (and we are not going to block staging).

I used one staging case to check this problem: https://openqa.suse.de/tests/10658338
Only the case which uses 'disable_boot_menu_timeout_right_key' can catch the problem, others cases will not be impacted even this PR merged.

jknphy commented:
> > Please, prepare also the staging test, we should reuse the textmode test suite that we recently added (coordinate with @rakoenig in https://progress.opensuse.org/issues/124676) Once addressed feedback and verified, we will merge the PR and resolve the ticket without enabling it in job groups for staging. We should file a ticket for the enablement (but test here that all works as expected) as the bug is still not fixed (and we are not going to block staging).
> 
> I used one staging case to check this problem: https://openqa.suse.de/tests/10658338 Only the case which uses 'disable_boot_menu_timeout_right_key' can catch the problem, others cases will not be impacted even this PR merged.

could you please file an additional WIP PR with the change on schedule for staging, so we remember to enable? (no ticket needed I guess then).

jknphy commented on schedule/staging/textmode_install_bsc1208266@64bit-staging.yaml:
```diff
@@ -0,0 +1,29 @@
+---
```
we don't need a new yaml schedule, we have already a textmode test suite in staging, please use that one.

tinawang123 replied:
If we use this yaml file, the staging case will be failed as bug has not fix yet.

jknphy replied:
> If we use this yaml file, the staging case will be failed as bug has not fix yet.

because we don't plan to merge the PR until the bug is fixed, it is just preparation.

tinawang123 commented on tests/installation/bootloader_settings/disable_boot_menu_timeout.pm:
```diff
@@ -12,6 +12,7 @@ use base 'y2_installbase';
 
 sub run {
     $testapi::distri->get_installation_settings()->access_booting_options();
+    $testapi::distri->get_bootloader_settings()->switch_to_bootloader_options_navigating_tabs();
```
Original code, tab to bootloader page and set grub timeout value in the same function 'sub disable_grub_timeout'.
I split this one function to two functions, one for switch to bootloader page, one for set grub timeout.
If we want to use right key to get bootloader page, we can reuse set grub timeout function.

jknphy replied:
then we need tow method:
`switch_to_bootloader_options` (doesn't need any prefix, because it is the only way to navigate with libyui-rest-api, going directly).
`switch_to_bootloader_options_via_tabs`

jknphy commented on tests/installation/bootloader_settings/disable_boot_menu_timeout_right_key.pm:
```diff
@@ -0,0 +1,20 @@
+
+sub run {
+    $testapi::distri->get_installation_settings()->access_booting_options();
+    send_key_until_needlematch 'inst-bootloader-options-highlighted', 'right';
```
for this case you need to wrap this in a method and put testapi in Controller, it is the exception of course.

jknphy commented on tests/installation/bootloader_settings/disable_boot_menu_timeout_right_key.pm:
```diff
@@ -0,0 +1,19 @@
+# Copyright SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Disable grub timeout from the Installer
```
I don't think nobody would understand what you mean with `_right_key` name, I would replace everywhere by `_navigating_tabs`. Do you want to test it with someone? wdyt?

jknphy commented on tests/installation/bootloader_settings/disable_boot_menu_timeout_right_key.pm:
```diff
@@ -0,0 +1,19 @@
+use strict;
+use warnings;
+use base 'y2_installbase';
+use testapi;
```
you don't need to import the library with using `$testapi`.

jknphy commented:
> Catch regression when sending right key when switching to tab Bootloader Options in ncurses
> 
> * Related ticket: https://progress.opensuse.org/issues/124610
> * Verification run: https://openqa.suse.de/tests/10912241#step/disable_boot_menu_timeout_right_key/1

and verification for textmode installation with minimal role in YaST job group ?

jknphy commented on lib/YaST/Bootloader/BootloaderSettingsController.pm:
```diff
@@ -79,6 +80,15 @@ sub write_to_partition {
     $self->get_boot_code_options_page->press_next();
 }
 
+sub disable_grub_timeout_right_key {
```
the existence of this function using testapi worth a mention to the bug and why is there in the form of a comment? wdyt?

jknphy commented on lib/YaST/Bootloader/BootloaderSettingsController.pm:
```diff
@@ -10,9 +10,11 @@ package YaST::Bootloader::BootloaderSettingsController;
 use YaST::Bootloader::BootCodeOptionsPage;
 use YaST::Bootloader::KernelParametersPage;
 use YaST::Bootloader::BootloaderOptionsPage;
+use YaST::workarounds;
```
it is not a workaround, the bug is fixed, it is just another testing path that we should cover to avoid regression, but we just need to mention that in a comment in the method.

tinawang123 commented:
Any more comment?

jknphy commented on tests/installation/bootloader_settings/disable_boot_menu_timeout_navigating_tabs.pm:
```diff
@@ -0,0 +1,18 @@
+# Copyright SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Disable grub timeout from the Installer accoding bsc#1208266
```
s/accoding/according/

jknphy commented on lib/YaST/Bootloader/BootloaderSettingsController.pm:
```diff
@@ -79,6 +88,13 @@ sub write_to_partition {
 
+sub disable_grub_timeout_navigating_tabs {
+    my ($self) = @_;
+    $self->get_bootloader_options_Tab->use_navigating_tabs();
```
your POM should represent some part of the screen, not just one Tab, plus should be in lower letter.
IMO the getter should be: `get_bootloader_settings_tabs()` (please, do not forget the `()` for each function to follow the current coding style).
And the function will be just `navigate()`,
so `$self->get_bootloader_settings_tab->navigate()`.

jknphy commented on lib/YaST/Bootloader/BootloaderOptionsTab.pm:
```diff
@@ -0,0 +1,34 @@
+# Copyright 2021 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: The class introduces all accessing methods for Bootloader Options tab
```
"bootloader options tab" is jut the one tab more on the right, it is BootloaderSettingsTabs the object to map.

jknphy replied:
it should have the name suggested above.
you are missing to create some `new()` method and you can drop the `init()`

jknphy commented on lib/YaST/Bootloader/BootloaderOptionsTab.pm:
```diff
@@ -0,0 +1,34 @@
+
+sub init {
+    my $self = shift;
+    $self->SUPER::init();
```
this class doesn't need to container NavigationBase or any libyui-rest-api,
to check that you are in the right screen, you just need existing method:
`get_boot_code_options_page`. I guess in the current design we would need another controller which would include `BootCodeOptionsPage` in similar way that `BootloaderSettingsController` does plus the needling code below.

tinawang123 replied:
I need this code: $self->{cmb_bootloader} = $self->{app}->combobox({id => "\"Bootloader::LoaderTypeWidget\""});
To make sure the page: https://openqa.suse.de/tests/10917817#step/disable_boot_menu_timeout_navigating_tabs/1 is shown
After this, I can use right key to get 'bootloader options' page

jknphy commented on lib/YaST/Bootloader/BootloaderSettingsController.pm:
```diff
@@ -79,6 +81,14 @@ sub write_to_partition {
+sub disable_grub_timeout_navigating_tabs {
+    my ($self) = @_;
+    $self->get_boot_code_options_page->is_shown();
+    YaST::Bootloader::BootloaderOptionsTab->navigate();
```
you need a getter, do not access directly to object

jknphy commented on lib/YaST/Bootloader/BootloaderSettingsController.pm:
```diff
@@ -79,6 +81,14 @@ sub write_to_partition {
 
+sub disable_grub_timeout_navigating_tabs {
+    my ($self) = @_;
+    $self->get_boot_code_options_page->is_shown();
```
I guess just `$self->get_boot_code_options_page()` you don't need to run `is_shown()` twice.

jknphy commented on lib/YaST/Bootloader/BootloaderSettingsController.pm:
```diff
@@ -10,9 +10,11 @@ package YaST::Bootloader::BootloaderSettingsController;
 use strict;
 use warnings;
 use YuiRestClient;
+use testapi;
```
now you can drop it.

jknphy commented on tests/installation/bootloader_settings/disable_boot_menu_timeout_navigating_tabs.pm:
```diff
@@ -0,0 +1,18 @@
+# Copyright SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Disable grub timeout from the Installer according bsc#1208266
```
instead of `according` perhaps would be more clear to phrase it like `to catch regression of`

tinawang123 commented:
Any more comments?

jknphy commented on lib/YaST/Bootloader/BootloaderOptionsTab.pm:
```diff
@@ -0,0 +1,26 @@
+}
+
+sub navigate {
+    my ($self) = @_;
```
not needed.

jknphy commented on lib/YaST/Bootloader/BootloaderSettingsController.pm:
```diff
@@ -79,6 +86,14 @@ sub write_to_partition {
 
+sub disable_grub_timeout_navigating_tabs {
+    my ($self) = @_;
+    $self->get_boot_code_options_page;
```
`$self->get_boot_code_options_page()`
